Layout: Implement innerText/outerText

Melchizedek6809 opened the pull request:
Hey, I've given the PR by @shanehandley over in #32427 the final polish, now it passes most of the Tests. I've had to diverge a bit from the algorithm as described in the whatwg spec since it only described a simplified version that would not give the same results as other browsers/WPT (it mainly left out a lot of special cases).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because WPT

shanehandley committed:
Implement outerText on HtmlElement

Signed-off-by: Shane Handley <shanehandley@fastmail.com>

Melchizedek6809 committed:
Fixed some innerText/outerText bugs

Signed-off-by: Benjamin Vincent Schulenburg <bennyschulenburg@gmx.de>

Wuelle commented on components/layout_2020/query.rs:
```diff
@@ -509,8 +514,415 @@ fn is_eligible_parent(fragment: &BoxFragment) -> bool {
+
+                let is_final_character_whitespace = transformed_text
+                    .chars()
+                    .last()
```
Minor thing, but `.chars().last()` is O(n) since it needs to to iterate over all the previous characters, you could use [`next_back()`](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html#tymethod.next_back) for O(1) instead.

Melchizedek6809 replied:
Ah thanks! I just changed it to use `next_back` instead.

Wuelle commented on components/layout_2020/query.rs:
```diff
@@ -509,8 +514,415 @@ fn is_eligible_parent(fragment: &BoxFragment) -> bool {
+                        let truncated_text = transformed_text
+                            .chars()
+                            .take(transformed_text.len() - 1)
+                            .collect();
```
[`String::pop`](https://doc.rust-lang.org/std/string/struct.String.html#method.pop) also removes the last character, but without (necessarily) allocating a new string. I think this could be used here?

Melchizedek6809 replied:
Nice, works great

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#dom-outertext>
+    fn GetOuterText(&self) -> Fallible<DOMString> {
```
I think it's worth adding a method for [`get-the-text-steps`](https://html.spec.whatwg.org/multipage/dom.html#get-the-text-steps), and use it here and within `GeInnerText`.

Melchizedek6809 replied:
Makes sense, they were treated separately until they hit Layout where both ended up calling the same method, I've added a `get_inner_outer_text` method to HTMLElement which is called by both `GetInnerText` and `GetOuterText`, I didn't want call this method `get_the_text_steps` since the actual implementation is in Layout, where the method is called just that.

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
+        Ok(DOMString::from(text))
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute%3Adom-outertext-2>
```
https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute:dom-outertext-2

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
+            fragment
+                .upcast::<Node>()
+                .AppendChild(text_node.upcast())
+                .unwrap();
```
We should either `expect` with the appropriate message if this cannot fail at this point, or return the error(even though the spec is silent on that).

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -467,52 +469,90 @@ impl HTMLElementMethods for HTMLElement {
-                    }
+        // Step 6: Replace this with fragment within this's parent.
+        if let Err(error) = parent.ReplaceChild(fragment.upcast(), node) {
+            warn!("Failed to replace child node: {:?}", error);
```
Here you can return the error, since the spec throws.

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
+        // of input.
+        let mut position = input.chars().peekable();
+
+        // Step : Let text be the empty string.
```
number

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
+        // Step 4
+        while let Some(ch) = position.next() {
+            match ch {
+                // While position is not past the end of input, and the code point at position is
```
We have some code at https://github.com/servo/servo/blob/ebed9218f2907767ba3c9dd9f27f30a6a6e9f225/components/net/fetch/headers.rs#L88

It might make sense to move it to `components/shared/net` and re-use it here.

Melchizedek6809 replied:
I fear then we'd have to iterate over the string twice since we also need to turn linefeeds into `<br>` elements and generate multiple Text nodes

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#merge-with-the-next-text-node>
+    fn merge_with_the_next_text_node(&self, node: DomRoot<Node>) {
```
We could make this faillible, instead of calling it inside the `node.is::<Text>()` condition(or you `expect`  below and explain why it cannot fail).

Melchizedek6809 replied:
Not sure about making this fallible, from the spec it doesn't sound like an operation that I'd expect to fail. It seems more like a clean up step to make sure that we don't have multiple Text nodes next to each other, which is already the case if either `node` or `next` are something other than Text nodes, or there is no `next`.

Maybe we should also call `unwrap` on the Result of the call to `ReplaceData`, since from what I can tell it only fails when fed invalid UTF-16, and I'm not sure about the architecture of Servo but invalid UTF-16 probably shouldn't make its way into the DOM at all, so we should probably fail as soon as possible in that case, or what do you think?

Melchizedek6809 replied:
I've tried to rewrite/clean it up a bit, what do you think about this version?

gterzian reviewed (changes requested):
Thanks for following the spec in such detail! Couple of points on the `script` side of things...

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
+
+        // If text is not the empty string, then append a new Text node whose data is text and node
+        // document is document to fragment.
+        if !text.is_empty() {
```
I'm confused about the loop, do you still need this here besides what is at [966](https://github.com/servo/servo/pull/33312/files#diff-978993aad649a5f5ef459e8071ed6b982d038e71ca6d53156e9dcdf74192c702R966)?

Melchizedek6809 replied:
Not sure whether the latest changes broke the line numbers, do you mean why there's the following code both inside the `while` loop and at the end?
```rust
if !text.is_empty() {
    append_text_node_to_fragment(&document, &fragment, text);
    text = String::new();
}
```

From what I understand that's because linefeeds should be converted to `<br>` tags, which means we might need to generate multiple text nodes. And since we might not end on a linefeed we have to put the accumulated characters into a final text node.

gterzian replied:
Ok, thanks for clarifying.

gterzian commented on components/script/dom/htmlelement.rs:
```diff
@@ -897,6 +937,82 @@ impl HTMLElement {
+                    return;
+                }
+
+                // Step 4: If next's parent is non-null, then remove next.
```
I don't see the parent condition in the spec, do you need it?

Melchizedek6809 replied:
Can probably be removed, since remove_self already makes sure there's a parent

Melchizedek6809 committed:
Unified innerText/outerText handling outside of Layout

Before these 2 were treated separately and only within
Layout would they end up calling the same method, now
they are already unified within HTMLElement

Signed-off-by: Benjamin Vincent Schulenburg <bennyschulenburg@gmx.de>

Melchizedek6809 commented:
@gterzian Thank you for the review! I still had some questions regarding some points in your review which I've added as comments, apart from that most things should be implemented now

mrobinson reviewed (commented):
I'll review this soon, but I have one question:

Does this need to be a layout query? It looks like it could simply run in the context of the script thread?

Melchizedek6809 commented:
Not sure if I understand correctly but as far as I can tell the code runs on the script thread with a couple of function calls in-between to abstract over the used layout engine.  As for why it's part of the layout_2020 crate: the code is making use of both `TextTransformation` and `WhitespaceCollapse` which seem to only be available in layout_2020, although I might be mistaken since I'm not familiar with the codebase.

mrobinson commented:
> Not sure if I understand correctly but as far as I can tell the code runs on the script thread with a couple of function calls in-between to abstract over the used layout engine. As for why it's part of the layout_2020 crate: the code is making use of both `TextTransformation` and `WhitespaceCollapse` which seem to only be available in layout_2020, although I might be mistaken since I'm not familiar with the codebase.

Apologies, I used the wrong terminology because layout and script used to run on separate threads. They don't any longer. What I mean is that this code is just iterating through DOM nodes and doesn't need access to laid out content. It does use layout's implementation of the text transformation rules, but those are utilities used in the process of layout. 

Do you foresee a further implementation of this API to need access to the laid out content of nodes? For instance, this might include knowing which nodes aren't included in a layout tree due to `display: none` or knowing about the details of `display: content`? If not, maybe it makes sense to move the implementation of this to the script trait so that it can be shared between both old and new layout and can be run without having to wait for a pending layout. We can think about where best to move the text transformation utilities.

Melchizedek6809 commented:
Ah, thank you for clearing that up!

We might need some additional information to properly apply `text-transform` on `:first-line` selectors, but other than that we don't need to concern ourselves too much with layout internals.

Apart from `script`, is there another crate we could move this implementation to? Just because it is already pretty big and slow to compile, so I'm a bit hesitant about moving more code in there than absolutely necessary.

Oh and do you know about the long term plans for layout_2013? I think I've read somewhere that there are plans for its removal now that layout_2020 is more mature. Just thinking out loud whether sharing the implementation between the 2 layout engines is worth the effort if the former is about to be removed anyways.

gterzian reviewed (approved):
Thanks for addressing the review, only some nits remains.

Melchizedek6809 commented on components/script/dom/htmlelement.rs:
```diff
@@ -897,6 +916,84 @@ impl HTMLElement {
+        }
+        // Step 3: Replace data with node, node's data's length, 0, and next's data.
+        let node_chars = node.downcast::<CharacterData>().unwrap();
+        let next_chars = next.downcast::<CharacterData>().unwrap();
```
Ah alright, I've changed these accordingly

Melchizedek6809 commented:
@gterzian Thanks for the review, those nits should now be resolved in with the newest commit.

mrobinson committed:
Address a few nits

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson reviewed (approved):
I think for now this can stay in the `layout` crate, as it may be that in the future this can be implemented against the display list instead of the DOM. Alternatively, it can be moved to `script` in a followup.

mrobinson commented:
Apologies for taking so long to review this one!

mrobinson commented:
There's a couple failures here:

```
OK /html/dom/elements/the-innertext-and-outertext-properties/getter.html

    FAIL [expected PASS] subtest: Whitespace around inline-flex should not be collapsed ("&lt;div&gt;abc &lt;span style='display:inline-flex'&gt;&lt;/span&gt; def")

    assert_equals: innerText expected "abc  def" but got "abc def"

FAIL [expected PASS] subtest: Whitespace around inline-grid should not be collapsed ("&lt;div&gt;abc &lt;span style='display:inline-grid'&gt;&lt;/span&gt; def")

assert_equals: innerText expected "abc  def" but got "abc def"

FAIL [expected PASS] subtest: Leading/trailing space removal at display:inline-flex boundary ("&lt;div&gt;123&lt;span style='display:inline-flex'&gt; abc &lt;/span&gt;def")

assert_equals: innerText expected "123abcdef" but got "123 abc def"

FAIL [expected PASS] subtest: Leading/trailing space removal at display:inline-grid boundary ("&lt;div&gt;123&lt;span style='display:inline-grid'&gt; abc &lt;/span&gt;def")

assert_equals: innerText expected "123abcdef" but got "123 abc def"
```

Melchizedek6809 commented:
> Apologies for taking so long to review this one!

No worries, thank you for the review!

> There's a couple failures here:

Alright, those seem to be the new WPT Tests I've added upstream (added a bunch in the process of writing this PR, since a couple of edge-cases weren't tested for before). Shouldn't take long to update this PR to make them pass.

Melchizedek6809 committed:
Added innerText support for `inline-flex`

Signed-off-by: Benjamin Vincent Schulenburg <bennyschulenburg@gmx.de>

Melchizedek6809 commented:
Alright, I've added support for inlineFlex and inlineGrid and updated the WPT results. Although the inlineGrid tests still fail because it seems that grid support is still enabled (when trying to enable it servo just paniced on running the WPT suite, I suppose grid is not quite stable yet in Servo). Once grid stabilizes and the display value will actually become `Display:InlineGrid` those 2 tests should pass.
